File: crates/version-spec/src/unresolved_spec.rs
use crate::resolved_spec::VersionSpec;
use crate::spec_error::SpecError;
use crate::syntax::*;
use crate::syntax_parser::parse_alias;
use compact_str::CompactString;
use human_sort::compare;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use std::fmt::{Debug, Display};
use std::str::FromStr;

/// Represents an unresolved version or alias that must be resolved
/// to a fully-qualified version.
#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(untagged, into = "String", try_from = "String")]
pub enum UnresolvedVersionSpec {
    /// A special canary target.
    Canary,
    /// An alias that is used as a map to a version.
    Alias(CompactString),
    /// A list of requirements.
    Range(Range),
    /// A partial version / requirement.
    Requirement(Requirement),
    /// A fully-qualified version.
    Version(Version),
}

impl UnresolvedVersionSpec {
    /// Parse the provided string into an unresolved specification.
    pub fn parse<T: AsRef<str>>(value: T) -> Result<Self, SpecError> {
        Self::from_str(value.as_ref())
    }

    /// Return true if the provided alias matches the current specification.
    pub fn is_alias<A: AsRef<str>>(&self, name: A) -> bool {
        match self {
            Self::Alias(alias) => alias == name.as_ref(),
            _ => false,
        }
    }

    /// Return true if the current specification is canary.
    pub fn is_canary(&self) -> bool {
        match self {
            Self::Canary => true,
            Self::Alias(alias) => alias == "canary",
            _ => false,
        }
    }

    /// Return true if the current specification can be treated as a
    /// fully qualified version, either calendar or semantic.
    pub fn is_fully_qualified(&self) -> bool {
        matches!(self, Self::Version(_))
    }

    /// Return true if the current specification is the "latest" alias.
    pub fn is_latest(&self) -> bool {
        match self {
            Self::Alias(alias) => alias == "latest",
            _ => false,
        }
    }

    /// Convert the current unresolved specification to a resolved specification.
    /// Note that this *does not* actually resolve or validate against a manifest,
    /// and instead simply constructs the [`VersionSpec`].
    ///
    /// Furthermore, the `Range` and `Requirement` variants will return a
    /// "latest" alias, as they are not resolved or valid versions.
    pub fn to_resolved_spec(&self) -> VersionSpec {
        match self {
            Self::Canary => VersionSpec::Canary,
            Self::Alias(alias) => VersionSpec::Alias(alias.to_owned()),
            Self::Version(version) => VersionSpec::Version(version.to_owned()),
            _ => VersionSpec::default(),
        }
    }

    /// Convert the current unresolved specification to a partial string, where
    /// minor and patch versions are omitted if not defined, and the comparator
    /// operator is also omitted. For example, "~1.2" would simply print "1.2".
    ///
    /// Furthermore, `Canary` will return "canary", `ReqAny` will return "latest",
    /// and aliases will return as-is.
    pub fn to_partial_string(&self) -> String {
        fn from_parts(major: u32, minor: Option<u32>, patch: Option<u32>, pre: &str) -> String {
            let mut version = format!("{major}");

            minor.inspect(|m| {
                version.push_str(&format!(".{m}"));
            });

            patch.inspect(|p| {
                version.push_str(&format!(".{p}"));
            });

            if !pre.is_empty() {
                version.push('-');
                version.push_str(pre);
            }

            version
        }

        match self {
            UnresolvedVersionSpec::Canary => "canary".into(),
            UnresolvedVersionSpec::Alias(alias) => alias.to_string(),
            UnresolvedVersionSpec::Range(_) => "latest".into(),
            UnresolvedVersionSpec::Requirement(req) => from_parts(
                req.major.unwrap_or_default(),
                req.minor,
                req.patch,
                req.prerelease.as_deref().unwrap_or_default(),
            ),
            UnresolvedVersionSpec::Version(ver) => {
                let version = from_parts(
                    ver.major,
                    Some(ver.minor),
                    Some(ver.patch),
                    ver.prerelease.as_deref().unwrap_or_default(),
                );

                match &ver.scope {
                    Some(scope) => format!("{scope}-{version}"),
                    None => version,
                }
            }
        }
    }
}

#[cfg(feature = "schematic")]
impl schematic::Schematic for UnresolvedVersionSpec {
    fn schema_name() -> Option<String> {
        Some("UnresolvedVersionSpec".into())
    }

    fn build_schema(mut schema: schematic::SchemaBuilder) -> schematic::Schema {
        schema.set_description("Represents an unresolved version or alias that must be resolved to a fully-qualified version.");
        schema.string_default()
    }
}

impl Default for UnresolvedVersionSpec {
    /// Returns a `latest` alias.
    fn default() -> Self {
        Self::Alias("latest".into())
    }
}

impl FromStr for UnresolvedVersionSpec {
    type Err = SpecError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        if value == "canary" {
            return Ok(UnresolvedVersionSpec::Canary);
        }

        // The grammar is the authority on what a version looks like, so try the
        // most specific shape first, and treat an alias as the residual: it is
        // whatever is not structurally a version, requirement, or range

        if let Ok(version) = Version::parse(value) {
            return Ok(Self::Version(version));
        }

        // Kept for the error below, as a failed requirement is the most
        // useful diagnostic for an input that was meant to be a version
        let error = match Requirement::parse(value) {
            Ok(req) => return Ok(Self::Requirement(req)),
            Err(error) => error,
        };

        if let Ok(range) = Range::parse(value) {
            return Ok(Self::Range(range));
        }

        match parse_alias(value) {
            Ok(alias) => Ok(Self::Alias(alias)),
            Err(_) => Err(error),
        }
    }
}

impl TryFrom<String> for UnresolvedVersionSpec {
    type Error = SpecError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        Self::from_str(&value)
    }
}

impl From<UnresolvedVersionSpec> for String {
    fn from(value: UnresolvedVersionSpec) -> Self {
        value.to_string()
    }
}

impl Display for UnresolvedVersionSpec {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Canary => write!(f, "canary"),
            Self::Alias(alias) => write!(f, "{alias}"),
            Self::Range(range) => write!(f, "{range}"),
            Self::Requirement(req) => write!(f, "{req}"),
            Self::Version(version) => write!(f, "{version}"),
        }
    }
}

impl PartialEq<VersionSpec> for UnresolvedVersionSpec {
    fn eq(&self, other: &VersionSpec) -> bool {
        match (self, other) {
            (Self::Canary, VersionSpec::Canary) => true,
            (Self::Canary, VersionSpec::Alias(a)) => a == "canary",
            (Self::Alias(a1), VersionSpec::Alias(a2)) => a1 == a2,
            (Self::Version(v1), VersionSpec::Version(v2)) => v1 == v2,
            _ => false,
        }
    }
}

impl AsRef<UnresolvedVersionSpec> for UnresolvedVersionSpec {
    fn as_ref(&self) -> &UnresolvedVersionSpec {
        self
    }
}

impl PartialOrd<UnresolvedVersionSpec> for UnresolvedVersionSpec {
    fn partial_cmp(&self, other: &UnresolvedVersionSpec) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for UnresolvedVersionSpec {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Self::Canary, Self::Canary) => Ordering::Equal,
            (Self::Alias(l), Self::Alias(r)) => l.cmp(r),
            (Self::Version(l), Self::Version(r)) => l.cmp(r),

            // Use human sorting for requirements/ranges
            _ => compare(&self.to_string(), &other.to_string()),
        }
    }
}

File: package/src/versions/unresolved-spec.ts
import { type Comparator, Range, SemVer } from "semver";
import { cleanVersionString, isAliasName } from "./utils";
import { VersionSpec } from "./resolved-spec";
import { PluginError } from "../errors";

/**
 * Represents an unresolved version or alias that must be resolved
 * to a fully-qualified and semantic result.
 */
export abstract class UnresolvedVersionSpec {
  /**
   * Convert the current unresolved specification to a resolved specification.
   * Note that this *does not* actually resolve or validate against a manifest,
   * and instead simply constructs the [`VersionSpec`].
   *
   * Furthermore, the `Req` and `ReqAny` variants will throw an error, as they
   * are not resolved or valid versions.
   */
  abstract toResolvedSpec(): VersionSpec;

  abstract toString(): string;

  toJSON() {
    return this.toString();
  }

  equals(other: { toString(): string }): boolean {
    return this.toString() === other.toString();
  }

  /**
   * Parse the provided string into an unresolved specification based
   * on the following rules, in order:
   *
   * - If the value "canary", map as `Canary` variant.
   * - If an alpha-numeric value that starts with a character, map as `Alias`.
   * - If contains `||`, split and parse each item with [`VersionReq`],
   *   and map as `ReqAny`.
   * - If contains `,` or ` ` (space), parse with [`VersionReq`], and map as `Req`.
   * - If starts with `=`, `^`, `~`, `>`, `<`, or `*`, parse with [`VersionReq`],
   *   and map as `Req`.
   * - Else parse with [`Version`], and map as `Version`.
   */
  static parse(value: string): UnresolvedVersionSpec {
    if (value == "canary") {
      return new UnresolvedVersionSpec.Canary();
    }

    value = cleanVersionString(value);

    if (isAliasName(value)) {
      return new UnresolvedVersionSpec.Alias(value);
    }

    // OR requirements (Node.js)
    if (value.includes("||")) {
      return new UnresolvedVersionSpec.ReqAny(value);
    }

    // AND requirements
    if (value.includes(",")) {
      return new UnresolvedVersionSpec.Req(value);
    }

    if (["=", "^", "~", ">", "<", "*"].includes(value.charAt(0))) {
      return new UnresolvedVersionSpec.Req(value);
    }

    const dotCount = [...value].reduce(
      (dots, char) => (char === "." ? dots + 1 : dots),
      0
    );

    if (dotCount < 2) {
      return new UnresolvedVersionSpec.Req(`~${value}`);
    }

    return new UnresolvedVersionSpec.Version(value);
  }
}

export namespace UnresolvedVersionSpec {
  /**
   * An alias that is used as a map to a version.
   */
  export class Alias extends UnresolvedVersionSpec {
    constructor(public readonly alias: string) {
      super();
    }

    override toResolvedSpec(): VersionSpec {
      return new VersionSpec.Alias(this.alias);
    }

    override toString(): string {
      return this.alias;
    }
  }

  /**
   * A special canary target.
   */
  export class Canary extends Alias {
    constructor() {
      super("canary");
    }

    override toResolvedSpec(): VersionSpec {
      return new VersionSpec.Canary();
    }
  }

  /**
   * A fully-qualified semantic version.
   */
  export class Version extends UnresolvedVersionSpec {
    public readonly version: SemVer;

    constructor(version: SemVer | string) {
      super();

      this.version = new SemVer(version);
    }

    override toResolvedSpec(): VersionSpec {
      return new VersionSpec.Version(this.version);
    }

    override toString(): string {
      return this.version.toString();
    }
  }

  /**
   * A list of requirements to match any against (joined by `||`).
   */
  export class ReqAny extends UnresolvedVersionSpec {
    public readonly range: Range;

    constructor(range: Range | string) {
      super();

      // This is pretty nasty, but we presort the ranges and comparators,
      // so the equality check is just a string comparison.
      const unsortedRange = new Range(range);
      const sortedRaw = unsortedRange.set
        .map((comparators) => [...comparators].sort().join(" "))
        .sort()
        .join("||");

      this.range = new Range(sortedRaw);
    }

    override toResolvedSpec(): VersionSpec {
      throw new PluginError("invalid operation");
    }

    override toString(): string {
      return this.range.range;
    }
  }

  /**
   * A partial version, requirement, or range (`^`, `~`, etc).
   */
  export class Req extends ReqAny {
    get comparators() {
      return this.range.set[0];
    }

    constructor(...comparators: readonly (Comparator | string)[]) {
      super(comparators.map((c) => c.toString()).join(" "));
    }
  }
}
